ddr/lib/ddr-ir/Macro.cpp

```diff
@@ -25,57 +25,951 @@
 #include "ddr/ir/Macro.hpp"
 
 #include "ddr/std/sstream.hpp"
+#include <ctype.h>
 #include <string.h>
+#include <stdint.h>
+#include <algorithm>
+#include <vector>
 
 using std::string;
 using std::stringstream;
 
-DDR_RC
-Macro::getNumeric(long long *ret)
+enum Symbol {
+	OR_OR,		/* || */
+	AND_AND,	/* && */
+	OR,			/* | */
+	XOR,		/* ^ */
+	AND,		/* & */
+	EQ_EQ,		/* == */
+	NOT_EQ,		/* != */
+	GT,			/* > */
+	LT,			/* < */
+	GT_EQ,		/* >= */
+	LT_EQ,		/* <= */
+	GT_GT,		/* >> */
+	LT_LT,		/* << */
+	PLUS,		/* + */
+	MINUS,		/* - */
+	MULT,		/* * */
+	DIV,		/* / */
+	REM,		/* % */
+	NOT,		/* ! */
+	TILDE,		/* ~ */
+	LPAREN,		/* ( */
+	RPAREN		/* ) */
+};
+
+class MacroScanner
 {
-	DDR_RC rc = DDR_RC_OK;
-	string value = getValue();
+private:
+	char const *_cursor;
+	char const *_end;
 
-	if (value.length() > 0) {
-		/* If the pre-processed macro contains no brackets, read it as a number. */
-		if (string::npos == value.find('(')) {
-			stringstream ss;
-			ss << value;
-			long long valueNumeric = 0;
-			ss >> valueNumeric;
-			if (ss.fail()) {
-				rc = DDR_RC_ERROR;
-			} else if (NULL != ret) {
-				*ret = valueNumeric;
-			}
+	MacroScanner(string expression) :
+		_cursor(expression.c_str()),
+		_end(_cursor + expression.length())
+	{
+	}
+
+	bool atSymbol(Symbol sym);
+	bool readNumber(int64_t *ret);
+
+	friend class MacroParser;
+};
+
+class MacroParser
+{
+/* data members */
+private:
+	MacroScanner _scanner; /*handles reading characters from expression string*/
+
+/* function members */
+private:
+	bool logicalOr(int64_t *ret);
+	bool logicalAnd(int64_t *ret);
+	bool bitwiseOr(int64_t *ret);
+	bool bitwiseXor(int64_t *ret);
+	bool bitwiseAnd(int64_t *ret);
+	bool equalNeq(int64_t *ret);
+	bool comparison(int64_t *ret);
+	bool bitshift(int64_t *ret);
+	bool addSub(int64_t *ret);
+	bool multDivRem(int64_t *ret);
+	bool unaryExpression(int64_t *ret);
+	bool parentheses(int64_t *ret);
+
+public:
+	MacroParser(std::string expression) :
+		_scanner(expression)
+	{
+	}
+
+	bool evaluate(int64_t *ret);
+};
+
+/**
+ * Helper function to read the value of a number from a string, this checks if
+ * a character is a digit in a specified base
+ *
+ * @param[in] c: character being checked
+ * @param[in] base: the base used to check if the char is valid
+ *
+ * @return: if the character is a valid digit in the specified base
+ */
+static bool
+isadigit(char c, int base)
```

Comment:
This function (and 'digitvalue') should be private static methods of MacroScanner.